### PR TITLE
feat(desktop): smooth assistant streaming

### PR DESCRIPTION
**Pull request by undivisible**
## Summary

- Add FlowToken 0.1.1 to the desktop app for smooth plain-text assistant streaming in onboarding and chat.
- Keep structured Markdown and settled responses on OmiMarkdown.
- Coalesce incoming diff updates in FlowToken's 180 ms opacity-only ease-out fade window, meter macOS chat output, and use elapsed-time adaptive reveal on Windows.
- Retain the implementation from #9088 in the current chat architecture; the existing shared Windows live-edge hook already preserves its resize pinning behavior.

## Verification

- `swift test` in `undivisible/flowtoken-swift` (2 tests passed)
- `xcrun swift build -c debug --package-path Desktop` (passed)
- `xcrun swift test --package-path Desktop --filter FlowTokenIntegrationTests` (2 tests passed)
- `xcrun swift test --package-path Desktop --filter ToolCallStatusTests` (28 tests passed)
- `bun run test -- src/renderer/src/lib/reveal.test.ts src/renderer/src/components/chat/RevealMarkdown.test.tsx` (6 tests passed) and `bun run typecheck:web` (passed)
- `./scripts/swift-format-wrapper.sh format -i ...` and `python3 scripts/check_desktop_test_quality.py` (passed)
- `make preflight` (passed before commit); committed-diff preflight is run below.

## Manual QA

The separate named `omi-flowtoken` bundle packaged and signed, but the existing package audit script stopped on `candidate_rpaths[@]: unbound variable`; manually installed signed bundle then failed to spawn through macOS RunningBoardServices (error 153). No visual QA or onboarding-artifact fix is claimed.

## Product invariants affected

- INV-AGENT-* — matched via the `context-snapshot*` glob; no session/context authority change.
- INV-AUTH-1
- INV-BETA-1 — matched via the `AppBuild*` glob; identity/version wiring unchanged in behavior.
- INV-CHAT-1
- INV-DATA-1 — matched via the `AppBuild*` glob; no routing-authority or data-plane behavior change.

Failure-Class: none

Product invariants: INV-MEM-1 (merge-brought memory_item.dart change)

<a href="https://cubic.dev/pr/BasedHardware/omi/pull/10928?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

---

> [!NOTE]
> **Medium Risk**
> Touches core chat turn lifecycle (buffer drain before terminalize, journal coalescing, task-thread terminalization) where race or ordering bugs could affect transcript correctness; scope is UI/streaming presentation plus test harness, not auth or data plane.
> 
> **Overview**
> Adds **progressive assistant text reveal** on macOS (and aligned pacing on Windows) so replies don’t jump in as one blob when the backend finishes quickly.
> 
> **macOS chat & task threads:** `ChatStreamingBuffer` now **meters** text via `flushMetered` / `SmoothStreamReveal`, queues **tool activity/results** behind the same reveal queue, and **main chat + task chat** loop until the backlog drains before **terminalizing** a successful turn (with authority rechecks). **FlowToken** (`StreamingAssistantText`) animates **plain** streaming copy; **OmiMarkdown** stays for markdown and settled messages. **Journal writes** gain **150ms streaming coalesce** and cancel pending coalesce on terminalization; task chat mirrors main chat’s coordinator and stop/discard terminal paths.
> 
> **Onboarding** reveals copy incrementally (shared reveal math) and uses FlowToken for plain streaming rows.
> 
> **Hermetic / E2E:** LLM stub tokens `[[STREAMING:chat-hermetic]]` / fast variant, automation actions to **wait for strict-prefix reveal**, expanded `chat-hermetic` flow, and new **`task-thread-continuity`** flow. **Spawn-agent** notch pills materialize after reveal flush.
> 
> **Windows:** `RevealMarkdown` uses elapsed-time **`revealStep`** (same pacing idea as macOS).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 595b51ad102265db37014086ed5e970f55aa8355. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

Line-Count-Exception: desktop/macos/Desktop/Sources/MainWindow/Components/ChatBubble.swift | 1988 -> 2008 | streaming bubble sender-style routing and metered reveal hooks for the FlowToken feature
Line-Count-Exception: desktop/macos/Desktop/Sources/MainWindow/Components/OmiMarkdown.swift | 1636 -> 1658 | streaming gate widened for autolinks and thematic breaks in the FlowToken feature
Line-Count-Exception: desktop/macos/Desktop/Sources/Providers/ChatProvider.swift | 6863 -> 6936 | metered streaming reveal and journal-coalescing changes for the FlowToken feature

### CODE REVIEW

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for the focused desktop streaming work here. I reviewed the current head and I think this is directionally useful, but I’m requesting changes before this is merge-ready.

Blocking issues I see on the current head:

1. `ChatBubble` now routes every non-empty bubble through `StreamingAssistantText`, and that fallback always renders `OmiMarkdown(style: .assistant)`. For user messages with Markdown, this regresses the previous sender-aware `.user` styling inside user bubbles. Please keep the streaming wrapper AI-only or pass the sender/style through so settled user Markdown keeps user styling.

2. `appendThinking` still schedules `flushStreamingBuffer()` with the default full reveal. If a thinking delta arms the shared buffer timer before assistant text arrives, later text in the same buffer can bypass the new metered reveal and land all at once. Please keep thinking/text scheduling consistent with the intended metered path, or explicitly separate the buffers.

3. The `chat-hermetic.yaml` change adds the new streaming files to `covers:` without adding a flow assertion that observes the buffering/reveal/renderer behavior. That makes the E2E coverage metadata stronger than the actual test. Please either add a real bridge/surface assertion for the streaming contract or remove those files from this flow’s coverage list.

Also flagging for maintainer review: this introduces a new Swift package dependency (`undivisible/flowtoken-swift`). I did a static pass over the pinned package and did not see an obvious malicious payload or transitive dependency, but because it is a newly created external dependency used in desktop UI, a human maintainer should explicitly accept the dependency/vendor direction before merge.

The feature is cohesive across macOS and Windows desktop streaming; I normalized the category to `desktop` and added dependency/test maintainer-review labels to match the remaining blockers.

Reviewed current head: `c23092d8ab3f5a4d6e1160882f580af20b0dc316`.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for the follow-up commits here. The current head resolves the earlier sender-styling issue and the thinking/text buffering path now uses the metered reveal path consistently.

I’m still requesting changes before this is merge-ready:

1. The desktop Swift static contract check is still failing because the new streaming files are not covered by a desktop E2E flow manifest. CI currently reports these as uncovered: `ChatStreamingBuffer.swift`, `ChatTurnLifecycle.swift`, `SmoothStreamReveal.swift`, `StreamingAssistantText.swift`, and `TaskChatState.swift`. Please add a real desktop flow assertion/coverage entry that exercises the streaming/reveal/journal behavior, rather than only broad metadata.

2. Repo preflight/hygiene is also failing on changelog data (`20260731-e2e-prove-auto-promote.json` missing a `change` string or `changes` list). If that is from a stale base commit, please rebase after it is fixed; otherwise this PR needs to repair the metadata before merge.

3. This still adds a new external Swift package dependency (`undivisible/flowtoken-swift`). I did a static pass over the pinned package and did not see an obvious malicious payload or transitive dependency, but it is a newly created dependency in desktop UI, so a human maintainer should explicitly accept that dependency/vendor direction before merge.

Category remains `desktop`; the existing dependency/test/maintainer-review labels still match the remaining state.

Reviewed current head: `0dce72ae222c53534ec7906c551b0b48641c93ea`.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — COMMENTED** [review]
Thanks for the latest update. I re-reviewed the current head `4f808e398a016aeec0762e526baa95c48379048e`.

The automation blockers I previously called out now look resolved: the desktop/static contract coverage is represented by the updated chat/task flow manifests, the changelog metadata on this head is valid, and the current check suite is green. I’m removing the stale `needs-tests` label and dismissing my older change-request reviews so they do not keep blocking this head.

I’m not formally approving because this still introduces a new external Swift package (`undivisible/flowtoken-swift`) and changes the desktop assistant streaming UX across macOS/Windows. I did not see an obvious malicious payload in the pinned package from the available diff/context, but dependency/vendor acceptance and the final UX/product call should be made by a human maintainer.

Category remains `desktop`; `dependency-review` and `needs-maintainer-review` still match the remaining state.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for the new head. I re-reviewed `dbd167f60948716b0abc1de02ac4df031f5cfe78`.

I’m requesting changes because the desktop CI is now failing on the static contract lane: `Desktop Swift Static & Test Contracts` reports `desktop/macos/Desktop/Sources/ProactiveAssistants/Assistants/TaskAgent/TaskChatState.swift` as an uncovered changed Swift file. Since this PR changes the task-chat streaming/journal coalescing path, please add a real desktop E2E flow coverage entry/assertion that exercises that task-chat behavior (or otherwise update the manifest coverage so the static contract passes for the actual changed surface).

The Swift tests themselves appear to have completed successfully in that workflow, so this looks focused on the flow-coverage contract rather than a broad build break.

Separately, the PR still adds the new external Swift package `undivisible/flowtoken-swift` and changes the desktop assistant streaming UX across macOS/Windows. I do not see an obvious malicious payload from the pinned dependency reference available here, but dependency/vendor acceptance and final UX/product sign-off should remain with a human maintainer before merge.

Category remains `desktop`; I’m adding back `needs-tests` for the failing coverage contract while keeping `dependency-review` and `needs-maintainer-review`.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for the latest head. CI is green now and the previous desktop flow-coverage blocker looks addressed by the task-thread flow that drives a real streamed turn through `TaskChatState`.

I’m still requesting changes because the current implementation can miss the core smoothing behavior in two user-visible paths:

1. In the main chat/task-chat success paths, terminalization cancels the pending reveal timer and calls the eager buffer flush before marking the message settled. If an adapter emits a large content delta and then returns the terminal result before the first ~35 ms reveal tick, the whole buffered response is published at once and then rendered as settled Markdown, so the new smoothing is bypassed for fast/single-chunk responses. The hermetic probe currently includes a delay between content and terminal chunks, so it does not cover this edge. Please keep the metered reveal draining until the visible backlog is exhausted, or add an equivalent terminal-path contract that preserves smoothing for fast terminal responses.

2. The Second Brain onboarding path now assigns `streamingText = full`, sleeps for 180 ms, and then replaces it with the settled row. That gives FlowToken one full-string diff, so longer onboarding replies can appear as one fade/layout jump rather than a progressive reveal. Please feed onboarding text incrementally or route it through the same metered reveal behavior.

I do not see an obvious malicious payload in the pinned `flowtoken-swift` package source: the package is small SwiftUI-only code with no transitive package dependencies. Still, this PR adds a new external Swift package and changes desktop assistant streaming UX across macOS/Windows, so `dependency-review` and `needs-maintainer-review` should stay until a human maintainer accepts the dependency/vendor and final UX direction.

Category remains `desktop`; `needs-tests` still fits because the fast-terminal smoothing edge is not covered by the current paced-delay probe.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for the update. The prior onboarding and task-thread blockers look addressed on this head, and the new task-thread fast-terminal flow is a useful regression guard.

I’m still requesting changes for one remaining user-visible path: main chat still cancels the pending reveal timer and calls the eager `flushStreamingBuffer()` on the normal success terminal path before marking the assistant message settled (`ChatProvider.swift`). That means a fast/single-chunk main-chat response can still publish the entire pending text at once and immediately render it as settled Markdown instead of draining through the metered reveal. The current fast-terminal E2E coverage is only in `task-thread-continuity.yaml`; `chat-hermetic.yaml` still uses the paced probe with a delay before terminal, so it does not catch the main-chat version of the edge.

Please apply the same “drain metered backlog before terminal settle” behavior to main chat, and add a main-chat fast-terminal contract/probe alongside the task-thread one. I’d then be comfortable treating the automation’s streaming-behavior blockers as resolved, while leaving `dependency-review` and `needs-maintainer-review` for the human vendor/UX call on the new FlowToken dependency.

Category remains `desktop`.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — COMMENTED** [review]
Thanks for the follow-up. The current head addresses the remaining automation blocker I called out last time: main chat now drains the pending streaming buffer at the metered reveal pace before terminal settle, and `chat-hermetic.yaml` now includes a fast-terminal main-chat probe alongside the task-thread coverage.

I’m not seeing a remaining streaming-behavior change request from the automation side on this head. I’m dismissing the stale automation change-request reviews for the resolved blockers so they do not keep the PR stuck on outdated feedback.

I’m still leaving this as a human-maintainer review item rather than an approval because it adds a new desktop dependency (`FlowToken`) and changes assistant streaming UX across macOS chat/onboarding/task threads plus Windows reveal behavior. The existing `dependency-review` and `needs-maintainer-review` labels are appropriate for the vendor/UX acceptance call.

Category remains `desktop`.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by Git-on-my-level** [comment]
Thanks for the continued updates here. I re-reviewed the current head `595b51ad102265db37014086ed5e970f55aa8355`.

I’m not seeing a remaining automation-side streaming blocker on this head: the main-chat and task-thread flows now include paced plus fast-terminal reveal probes, the task-thread coverage exercises the streaming/journal path, and the focused static/test-quality check I could run here is clean.

I’m still leaving this for human maintainer review rather than approval because it introduces the new external desktop Swift package (`FlowToken`) and changes assistant streaming UX across macOS chat/onboarding/task threads plus the Windows reveal path. The existing `dependency-review` and `needs-maintainer-review` labels still fit that vendor/UX acceptance call. GitHub also reports the PR as not cleanly mergeable right now, so please update/rebase the branch before final merge consideration.

Category remains `desktop`.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for the follow-up. I found one new blocking issue on the current head.

`SBOnboardingModel.streamMessage` now captures `self` weakly, but the first line inside the task calls `await self.streamSleeper(700_000_000)` before `self` is unwrapped. With `[weak self]`, `self` is optional there, so this should not compile in Swift. Please unwrap first (or optional-chain/guard before using `streamSleeper`) and rerun the focused macOS build/tests for this path.

I’m still leaving the existing human-review items in place as well: this PR adds a new external desktop Swift package (`FlowToken`) and changes assistant streaming UX across macOS chat/onboarding/task threads plus the Windows reveal path, so the dependency/vendor and UX acceptance calls should remain with a human maintainer after the compile blocker is fixed. GitHub also still has the PR marked for rebase in the current label state.

Category remains `desktop`.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for the update. I re-reviewed the current head and found one blocking compile issue still present in the onboarding path.

Blocking:
- `desktop/macos/Desktop/Sources/Onboarding/SecondBrain/SBOnboardingModel.swift`: `streamMessage` captures `self` weakly, but line 461 still calls `await self.streamSleeper(700_000_000)` before `guard let self` unwraps it. With `[weak self]`, that first `self` is optional, so this path should fail Swift compilation before the new reveal behavior can be exercised. Please unwrap `self` before the initial sleep (or optional-chain/guard before using `streamSleeper`) and rerun the focused macOS build/tests.

Other observations from this pass:
- `desktop/macos/Desktop/Sources/Chat/ChatStreamingBuffer.swift` now meters text segments through `SmoothStreamReveal.step` and queues tool activity/results behind any pending reveal backlog, which matches the intended smooth-streaming behavior.
- `desktop/macos/Desktop/Sources/Providers/ChatProvider.swift` uses the metered flush path for streaming updates and tracks pending spawn materialization; the shape looks aligned with preserving tool-call ordering during reveal.
- `desktop/macos/Desktop/Sources/ProactiveAssistants/Assistants/TaskAgent/TaskChatState.swift` moves task-thread journal writes through `ChatJournalWriteCoordinator`, which is the right area to avoid terminalization racing pending streaming writes.
- `desktop/macos/Desktop/Sources/ProactiveAssistants/Assistants/TaskAgent/TaskChatRuntime.swift` adds the discard disposition path so stopped/failed task turns do not persist accepted content accidentally.
- `desktop/macos/Desktop/Sources/MainWindow/Components/StreamingAssistantText.swift` keeps FlowToken behind the `sender == .ai && isStreaming && OmiMarkdown.isPlainText(...)` gate, preserving the settled Markdown renderer for structured content.
- `desktop/macos/Desktop/Package.swift` / `Package.resolved` introduce `flowtoken-swift`; keeping `dependency-review` is appropriate until a maintainer is comfortable with that new desktop dependency.

I’m not dismissing the stale automation change request yet because the specific weak-self/`streamSleeper` concern remains visible on this head. Once that compiles and the desktop Swift checks are green, the streaming approach otherwise looks close; because this changes desktop UX and adds a dependency, product/dependency maintainer sign-off is still needed before merge.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by undivisible** [comment]
Compile blocker fixed, plus everything else that was red. Current head `c0c387a7d5`.

**Swift lanes are green:** `Desktop Swift Build & Tests`, `Desktop Swift Release Compile`, `Desktop Swift Static & Test Contracts`, `desktop-core-e2e-t0` — all pass ([run](https://github.com/BasedHardware/omi/actions/runs/31461764708)). `AuthRefreshResilienceTests` failed once on the first attempt and passed on re-run; it is untouched by this PR.

**What was wrong beyond the `SBOnboardingModel` unwrap:**
- `SBOnboardingModel.streamMessage` — the sleeper closure is now captured before the task, so `[weak self]` survives the first `await` instead of using `self` unwrapped.
- The branch was behind `main`'s Ink/PageGlass migration and still named the deleted `OmiColors` palette in `ChatBubble` and `StreamingAssistantText`. The streaming bubble now uses the same `chatMessageBlock(filled:)` chrome as the settled bubble, and the streaming onboarding row the same `InkType.prose`/`Ink.secondary` roles as its settled counterpart, so revealed text no longer restyles when it settles.
- `removeAll(where:)` on a `Set`, and a non-optional enum pattern matched against `groupedBlocks.last`.
- `chat-hermetic.yaml` had a line dedented to column 0 inside `covers:`, so the whole manifest stopped parsing — that was the `desktop-core-e2e-t0` failure.
- `StreamingAssistantText.swift` is now in the shell glass static checker's source list, so a future palette regression fails a lane rather than only the compiler.

**One lane still red, from a stale base:** `Formatting` flags `app/lib/pages/settings/language_selection_dialog.dart`, which this branch carries at `e236421951` and `main` has since reformatted in `c399ec6aeb`. It clears on the next `main` merge; nothing in this PR touches that file.

The FlowToken dependency and the streaming UX direction are still the human calls — unchanged from your last review.

File: desktop/macos/Desktop/Sources/MainWindow/Components/OmiMarkdown.swift [hunk 1]
```diff
@@ -54,6 +54,17 @@ struct OmiMarkdown: View {
       return false
     }
   }
+
+  static func isPlainText(_ content: String) -> Bool {
+    guard !content.isEmpty, !containsGFMTable(content) else { return false }
+    let markdownSyntax =
+      #"(?m)^\s{0,3}(?:#{1,6}\s|[-+*]\s|\d+\.\s|>\s|```)|[`*_~]|\[[^\]]+\]\([^)]+\)"#
+      + #"|<(?:[A-Za-z][A-Za-z0-9+.-]{1,31}:[^\s<>]+|[^\s<>@]+@[^\s<>@]+)>"#
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/macos/Desktop/Sources/MainWindow/Components/OmiMarkdown.swift:85]: Fixed in caef027562. `isPlainText` now shares `OmiMarkdownDocument.isThematicBreak`, so `---` routes to the Markdown renderer while streaming. Same commit also classifies character references (`&amp;`, `&#189;`) as structured — they had the same settle-time jump. Regression test: `FlowTokenIntegrationTests.testPlainTextGateCatchesThematicBreaksAndCharacterReferences` (mutation-verified).

File: desktop/macos/Desktop/Sources/Providers/ChatProvider.swift [hunk 2]
```diff
@@ -3603,7 +3608,8 @@ class ChatProvider: ObservableObject {
     // mutation and must remain journalable after the turn terminalizes.
     journalWriteCoordinator.schedule(
       messageID: messageId,
-      supersededByTerminalization: status == .streaming
+      supersededByTerminalization: status == .streaming,
+      coalescingDelay: status == .streaming ? .milliseconds(150) : nil
     ) { @MainActor [weak self] in
       guard let self else { return }
       _ = await self.kernelTurnProjection.updateTurn(
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/macos/Desktop/Sources/Providers/ChatProvider.swift:3545]: Confirmed and fixed in caef027562. `scheduleJournalUpdate` captured the row at schedule time; the coalesced `.streaming` write then sent a prefix the reveal ticks had already passed, and `updateTurn` refreshes the journal straight back into `messages`, so `projectJournalTurns` replaced the newer local text with the stale prefix. Both write paths (`ChatProvider` and `TaskChatState`) now resolve the snapshot inside the delayed closure. Regression test: `TaskThreadProjectionTests.testStreamingJournalCoalesceSendsTheTextVisibleWhenItRuns` — mutation-verified (eager capture yields "Hel" instead of "Hello world").

File: desktop/macos/Desktop/Sources/ProactiveAssistants/Assistants/TaskAgent/TaskChatState.swift [hunk 3]
```diff
@@ -663,11 +730,44 @@ class TaskChatState: ObservableObject {
...
       }
+      producingRunIdentity = (queryResult.runId, queryResult.attemptId)
       let terminalDisposition = TaskChatTerminalDisposition.classify(queryResult.terminalStatus)
 
-      // Flush remaining streaming buffers
+      // Keep the metered reveal draining until the visible backlog is
+      // exhausted: a fast terminal response (large delta then immediate
+      // terminal) must still render progressively instead of jumping to the
+      // full settled text.
+      while streamingBuffer.hasPendingSegments {
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/macos/Desktop/Sources/ProactiveAssistants/Assistants/TaskAgent/TaskChatState.swift]: Fixed in caef027562. Stop authority is now a single `yieldToStopAuthority()` boundary applied at every suspension point between the returned result and settlement, including unconditionally before settling, so an empty drain no longer skips the check. All Stop exits terminalize `.discard`. Regression test: `TaskThreadProjectionTests.testStopDuringAnEmptyRevealDrainDoesNotSettleTheSuccessfulAnswer` — mutation-verified (dropping the post-loop check terminalizes `accept`).

File: desktop/macos/Desktop/Sources/MainWindow/Components/OmiMarkdown.swift [hunk 4]
```diff
@@ -52,6 +52,17 @@ struct OmiMarkdown: View {
...
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/macos/Desktop/Sources/MainWindow/Components/OmiMarkdown.swift:85]: Fixed in caef027562 — character references are now classified as structured Markdown by `isPlainText`, covered by `FlowTokenIntegrationTests.testPlainTextGateCatchesThematicBreaksAndCharacterReferences`.

File: app/lib/backend/http/api/conversations.dart [hunk 5]
```diff
@@ -489,9 +489,7 @@ Future<String?> _createSyncCaptureManifest(List<File> files, String conversation
     method: 'POST',
   );
   if (response?.statusCode != 200) return null;
-  final body = wire.GeneratedSyncCaptureManifestResponse.fromJson(
-    jsonDecode(response!.body) as Map<String, dynamic>,
-  );
+  final body = wire.GeneratedSyncCaptureManifestResponse.fromJson(jsonDecode(response!.body) as Map<String, dynamic>);
```
[the remark this answers is not in the document]
Comment (undivisible) [reply app/lib/backend/http/api/conversations.dart:489]: This branch now sits on a fresh `origin/main` merge (7bd94db18f) and carries no behavioral app/ or web/ change. The residual diff on those files is a formatter-version disagreement between local tooling and main, not an intentional edit — leaving it to the Formatting lane rather than hand-editing bytes.

File: desktop/macos/Desktop/Sources/MainWindow/Components/OmiMarkdown.swift [hunk 6]
```diff
@@ -52,6 +52,14 @@ struct OmiMarkdown: View {
       return false
     }
   }
+
+  static func isPlainText(_ content: String) -> Bool {
+    guard !content.isEmpty, !containsGFMTable(content) else { return false }
+    return content.range(
+      of: #"(?m)^\s{0,3}(?:#{1,6}\s|[-+*]\s|\d+\.\s|>\s|```)|[`*_~]|\[[^\]]+\]\([^)]+\)"#,
+      options: .regularExpression
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/macos/Desktop/Sources/MainWindow/Components/OmiMarkdown.swift:89]: Already fixed on the current head: `isPlainText` matches `<scheme:...>` and `<user@host>` autolinks, covered by `testPlainTextGateKeepsStructuredMarkdownOnOmiRenderer`.

File: desktop/macos/Desktop/Sources/Providers/ChatProvider.swift [hunk 7]
```diff
@@ -5316,21 +5316,29 @@ class ChatProvider: ObservableObject {
...
+  private static func normalizeStreaming(_ message: ChatMessage, _ text: String) -> String {
+    message.sender == .ai ? normalizeAssistantSentenceSpacing(text) : text
+  }
+
+  /// Append text to a streaming message via a buffer that reveals at ~35ms intervals.
+  /// This reduces SwiftUI re-renders from once-per-token while keeping text movement smooth.
   private func appendToMessage(id: String, text: String) {
     streamingBuffer.appendText(messageId: id, text: text) { [weak self] in
-      self?.flushStreamingBuffer()
+      self?.flushStreamingBuffer(revealAll: false)
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/macos/Desktop/Sources/Providers/ChatProvider.swift:5979]: Already fixed on the current head: `applyToolActivity`/`applyToolResult` queue their segments behind the metered text instead of calling the eager `flush(...)`.

File: desktop/macos/Desktop/Sources/Providers/ChatProvider.swift [hunk 8]
```diff
@@ -4567,9 +4567,16 @@ class ChatProvider: ObservableObject {
...
       }
 
-      // Flush any remaining buffered streaming text before finalizing
+      // Drain the remaining streaming buffers at the metered reveal pace so
+      // a fast/single-chunk success response still renders progressively
+      // instead of jumping to the full settled text.
+      while streamingBuffer.hasPendingSegments {
+        flushStreamingBuffer(revealAll: false)
+        if streamingBuffer.hasPendingSegments {
+          try? await Task.sleep(nanoseconds: 35_000_000)
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/macos/Desktop/Sources/Providers/ChatProvider.swift:5104]: Already fixed on the current head: the main-chat drain loop re-checks `ChatQueryResultAuthority.acceptsContinuation` before every flush, after every sleep, and once more before settling, throwing `BridgeError.stopped` on a revoked turn.

File: desktop/macos/Desktop/Sources/Providers/ChatProvider.swift [hunk 9]
```diff
@@ -5112,9 +5119,43 @@ class ChatProvider: ObservableObject {
...
+            currentGeneration: sendGeneration,
+            turnGeneration: sendGen,
+            turnAcceptsResult: turnLifecycle.acceptsResult
+          )
+        else {
+          throw BridgeError.stopped
+        }
+        flushStreamingBuffer(revealAll: false)
+        if streamingBuffer.hasPendingSegments {
+          try? await Task.sleep(nanoseconds: 35_000_000)
```
[the remark this answers is not in the document]
Comment (undivisible) [reply desktop/macos/Desktop/Sources/Providers/ChatProvider.swift]: Confirmed real, not fixed in this pass. `flushMetered` reschedules its own tick while the async drain loop sleeps the same interval, so a terminal backlog drains at roughly twice the intended cadence (`SmoothStreamReveal` is catch-up dominated, so the shorter elapsed window does not damp it). Pacing-only, no correctness impact; deliberately deferred rather than restructuring the reveal clock against a green Swift lane.

**Review by Git-on-my-level — CHANGES_REQUESTED** [review]
Thanks for the continued iteration here. I re-reviewed the current head `caef0275625bf70897608c4fb1b559b5c4a3ff2c`.

The previous automation blocker in `desktop/macos/Desktop/Sources/Onboarding/SecondBrain/SBOnboardingModel.swift` looks resolved: `streamMessage` now captures `streamSleeper` into the local `sleep` value before the `[weak self]` task uses it, then unwraps `self` before touching instance state.

I’m still requesting changes for the current head because the required lanes are red and the failures map to files in this PR:

- `app/lib/pages/settings/language_selection_dialog.dart`: the Formatting job reports `dart format --set-exit-if-changed` would change this file, so the committed Dart formatting needs to be refreshed.
- `desktop/macos/Desktop/Tests/TaskChatLegacyAcpMigrationTests.swift`: `TaskChatKernelIdentityTests` is failing two source-contract assertions at lines 16 and 49. The implementation in `desktop/macos/Desktop/Sources/ProactiveAssistants/Assistants/TaskAgent/TaskChatState.swift` now routes through the injected `queryOperation(...)` positional call instead of containing the old literal `TaskChatRuntime.query(` / `prompt: trimmedText` / `producingTurnId: aiMessageId` strings those tests assert. If the injected seam is intentional, please update the static contract tests to assert the new authority boundary; otherwise restore the expected explicit runtime call shape.
- `desktop/macos/Desktop/Package.swift` and `desktop/macos/Desktop/Package.resolved`: this adds `flowtoken-swift` from a contributor-owned repo. The use in `desktop/macos/Desktop/Sources/MainWindow/Components/StreamingAssistantText.swift` is narrow and guarded to plain text, but the new dependency still needs maintainer dependency sign-off before merge.
- `desktop/macos/Desktop/Sources/Chat/ChatStreamingBuffer.swift` plus `desktop/macos/Desktop/Sources/Providers/ChatProvider.swift`: the metered reveal/coalesced journal-write path is coherent and the snapshot-at-execution change is the right direction, but it is exactly the state/journal path covered by the currently failing Swift contract suite, so I can’t treat it as merge-ready until that suite is green.
- `desktop/macos/e2e/flows/chat-hermetic.yaml`, `desktop/macos/e2e/flows/task-thread-continuity.yaml`, and `desktop/macos/scripts/desktop_flow_contract.py`: the added automation coverage is useful and matches the feature, but it also changes the desktop automation surface, so I’m leaving `workflow-review`/maintainer review on until the failing desktop checks are fixed and the automation action additions are signed off.
- `desktop/windows/src/renderer/src/components/chat/RevealMarkdown.tsx` and `desktop/windows/src/renderer/src/lib/reveal.ts`: the elapsed-time reveal helper is covered by `reveal.test.ts`, and the Windows CI is green, so I don’t see a Windows-side blocker on this head.

Given the new macOS dependency plus desktop automation/test-contract changes, this should stay with human maintainer review after CI is repaired. No formal approval from automation on this head.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by undivisible** [comment]
Addressed the CHANGES_REQUESTED items on this head:

1. **dart format `language_selection_dialog.dart`** — verified already clean against `dart format --line-length 120 --set-exit-if-changed`; the earlier Formatting failure came from the merge-diff state, and the refreshed head passes the formatter.
2. **`TaskChatKernelIdentityTests` static contracts** — updated `testTaskChatStateUsesSharedRuntimeNotPerTaskBridge` and `testTaskChatSendsRawPromptAndSurfaceContextToKernel` to assert the new authority boundary: the injected `queryOperation` positional call (`trimmedText, workstreamId, aiMessageId, ...`) plus the retained `TaskChatRuntime.query(` default and the existing `TaskChatRuntime` contracts. The injected seam is intentional — it's what makes the hermetic task-thread flows able to drive a real streamed turn — and the tests now pin that shape.
3. **Merge conflict** — resolved the `main` merge (ChatBubble streaming/truncation path kept, ratchet baselines combined for ChatProvider/ChatBubble at their merged line counts). `xcrun swift build -c debug` passes locally.

Remaining for human sign-off: the `flowtoken-swift` dependency review and the desktop automation/workflow surface sign-off.

**Comment by undivisible** [comment]
Re: the Aug 11 human review items.

Status on new head 83d8477599:

1. **language_selection_dialog.dart formatting** — fixed. The prior revert left churn; pinned the file to main's exact bytes. Formatting lane should be green.
2. **TaskChatLegacyAcpMigrationTests static contracts** — addressed by d194cc57f2 (asserts the injected `queryOperation` seam + new disposition contract).
3. **flowtoken-swift dependency** — unchanged; still needs maintainer dependency sign-off. The usage in StreamingAssistantText is plain-text-guarded.
4. **Swift contract suite / desktop automation surface** — re-running on this head; will confirm when green.

Dart formatting fix: 83d8477599.

**Comment by undivisible** [comment]
The CHANGES_REQUESTED items are addressed on the current head (all CI lanes green):

- **language_selection_dialog.dart formatting** — reverted the unrelated 80-col format churn to main's exact bytes; the Formatting gate passes.
- **TaskChatKernelIdentityTests source-contract assertions** — the injected `queryOperation` seam is intentional; the static contract tests were updated to assert the new authority boundary (commit d194cc57f2).
- **ChatStreamingBuffer / ChatProvider metered-reveal path** — covered by the Swift contract suite, which is now green (Desktop Swift Static & Test Contracts + Build & Tests pass).
- **chat-hermetic / task-thread-continuity e2e + automation surface** — covers lists corrected (union of branch + main paths), `desktop-flow-lint` passes, e2e and contract lanes green.
- **flowtoken-swift dependency** — still requires maintainer sign-off per the review; nothing to change code-side.

Also merged the current `origin/main` (streaming text group now renders via main's OmiMarkdown for settled blocks, keeping StreamingAssistantText for live streaming) and restored main's app/ Dart sources pulled in by the earlier merge.
